pkgs/sops-install-secrets/main.go

```diff
@@ -116,13 +117,25 @@ func readManifest(path string) (*manifest, error) {
 	return &m, nil
 }
 
+func linksAreEqual(linkTarget, targetFile string, info os.FileInfo, secret *secret) bool {
+	validUG := true;
+	if stat, ok := info.Sys().(*syscall.Stat_t); ok {
+		validUG = validUG && int(stat.Uid) == secret.owner
+		validUG = validUG && int(stat.Gid) == secret.group
+	}
+	return linkTarget == targetFile && validUG
+}
+
 func symlinkSecret(targetFile string, secret *secret) error {
 	for {
 		stat, err := os.Lstat(secret.Path)
 		if os.IsNotExist(err) {
 			if err := os.Symlink(targetFile, secret.Path); err != nil {
 				return fmt.Errorf("Cannot create symlink '%s': %w", secret.Path, err)
 			}
+			if err := os.Lchown(secret.Path, secret.owner, secret.group); err != nil {
```

Comment:
The reason I wanted to use fchown is that a potential attacker could replace the symlink with a file and gaining access to a file that was not supposed to be read.

To circumvent that the following approach makes sure we only change the owner of our own symlink:

1. fd = open() with O_PATH set -> this allows to open symlinks
2. readlinkat(fd, ""); the "" is a special case which allows to operate on file descriptors
3. check if the result of readlinkat is as expected
4. fchown(fd, ...)

I have not checked if the go stdlib exposes all these but `golang.org/x/sys/unix` shoul

Reply:
I just want to make sure I understand. We create the symlink and then immediately `Lchown` it. It's the only time Lchown would be called. So the attacker in this scenario can pause execution, or time-attack fast enough to switch the symlink after we create, and before we symlink?

I do think I understand that the method you described will operate on the exact symlink-fd and let us ensure the target is valid before we act on it.

Since this always happens when we create the symlink, is it easier to create the symlink initially with the correct owner and group?

Reply:
Creating a symlink initially with the correct permission is possible too, but I am not sure it is easier. I think what would work is to use `setfsuid`/`setfsguid` to run the system call with the user's owner/group. However this can fail if the user does not have actually write access to the directory you are trying to write. TOCTOU attacks in the filesystem are certainly possible - I have implemented them myself - There are beginner examples in CTFs.

Reply:
At the end of the day do we want to allow users to interfere with system provisioned symlinks?